Fix ICON package (config wrapper) for spack v0.18.1

mjaehn opened the pull request:
This goes together with https://github.com/C2SM/icon-exclaim/tree/spack_v0.18.1

mjaehn commented:
launch jenkins icon

jonasjucker reviewed (changes requested):
You safed us ;)

Where do I find the modified config-wrapper? icon-exclami, icon-c2sm?

Also could you (at least temporary) include a version in the icon package that we can test here?
So that launch jenkins icon does something meaningful?

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins icon daint

mjaehn commented:
launch jenkins daint icon

mjaehn commented:
launch jenkins daint icon

mjaehn commented:
launch jenkins daint icon

mjaehn commented:
launch jenkins daint icon

PanicSheep commented:
launch jenkins daint icon

mjaehn commented:
Things done so far:

- [x] created spack environment files for different builders (`daint_gcc_cpu`, `daint_nvhpc_cpu`, `daint_nvhpc_gpu`) in [ICON repo](https://github.com/C2SM/icon-exclaim/tree/spack_v0.18.1)
- [x] fixed config files
- [x] added `cdo` and `python` dependencies to ICON package
- [x] added functions to do testing with devbuild within active spack environment
- [x] execute test script with installed `python` version

At the moment, the spack-c2sm system-tests are failing, even though the test script `./scripts/spack/test.py` works locally. The logfile doesn't really specify what is happening, so I would appreciate some help here.

mjaehn commented:
launch jenkins daint icon

PanicSheep commented on packages/icon/package.py:
```diff
@@ -61,7 +65,8 @@ class Icon(Package):
     depends_on('netcdf-c')
     depends_on('cuda', when='icon_target=gpu')
-    depends_on('cdo', type='test')
+    depends_on('cdo')
```
Why was it necessary to change cdo from a test dep to a build+link dep? (Deps are by default build+link)

jonasjucker replied:
If we execute the test.py script outside of spack, concretization does not include CDO in its spec.
But cdo is needed for all tests in ICON.

jonasjucker replied:
It also does not work when having cdo as test dep and calling the script from package.py of ICON.

PanicSheep commented on sysconfigs/daint/packages.yaml:
```diff
@@ -162,10 +162,6 @@ packages:
     externals:
     - spec: libtool@2.4.6
       prefix: /usr
-  libxml2:
```
Why was this removed?

jonasjucker replied:
There was a version incompatability when spack picked libxml2 from /usr.

PanicSheep reviewed (commented):
I refactored the 'spack_install_and_test' functions quite a bit in master, so there's going to be merge conflicts. Shall I do this merge now? I think I know how to resolve them pretty simple. I'd extend the existing function to conditionally take spack environments. Then all packages are able to make use of that feature.

jonasjucker commented:
Since we are back on track with the "pure" ICON package I consider this branch as a backup in case we really cannot make it work until the upgrade.
So far no action from your side here is needed.
